### PR TITLE
Tickets/dm 5745

### PR DESCRIPTION
**Pull request by afausti**
@jhoblitt Implemented API for listing and creating jobs, metrics and measurements in the dashboard app. 

For a given metric registered in the database, a job can use the API to post information like job name, build number, start time, duration, status and the value of the metric measurement.

A description can be found at http://sqr-009.lsst.io/en/latest/

### CODE REVIEW

File: squash/tests/test_api.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+"""Tests for the dashboard API"""
```
Comment (jhoblitt) [inline squash/tests/test_api.py]: This file doesn't have the execute bit set on it.

**Comment by jhoblitt** [comment]
@afausti I had no problem starting up a django instance.  Note that you have to be in the `qa-dashboard/squash` directory in order to execute `run.sh` as it is looking for `manage.py` in the CWD.  A small README with test instructions would be helpful but isn't critical.

I was able to run `test_post` from `test_api.py` and check that the value does land in the sqlite db.  I don't think this is yet rendered via the dashboard UI?

```
sqlite> select * from dashboard_metric;
1|PA1|Photometric Repeatability|millimag|<|8.0|5.0|3.0|5.0
```

I was also able to read this value out via a rest call (after some trial and error).  It wasn't immediately obvious to me that the administrative account name was that of the user that executed `run.sh`.

```
 curl -i --user jhoblitt:foo -H "Accept: application/json" -H "Content-Type: application/json" -X GET http://localhost:8000/api/metric/
HTTP/1.0 200 OK
Date: Wed, 20 Apr 2016 18:31:03 GMT
Server: WSGIServer/0.1 Python/2.7.11
Vary: Accept
X-Frame-Options: SAMEORIGIN
Content-Type: application/json
Allow: GET, POST, HEAD, OPTIONS

{"count":1,"next":null,"previous":null,"results":[{"metricId":1,"name":"PA1","description":"Photometric Repeatability","units":"millimag","condition":"<","minimum":8.0,"design":5.0,"stretch":3.0,"user":5.0,"links":{"self":"http://localhost:8000/api/metric/1/"}}]}
```

I'm afraid that I know almost nothing about django, so I'm not sure if I can give any useful feedback on the code.  This looks reasonable to my naive eye.  pylint complains about a few minor issues such as missing doc strings but I suspect it doesn't make sense to doc tag the ORM relationship classes.

File: squash/tests/test_api.py (hunk 1 above)
Comment (afausti) [inline squash/tests/test_api.py]: If I  run py.test in the command line the tests/test_api.py is found and the tests are executed, but it requires the application to be running which is not good. I probably should switch to django test framework then a database would be created for the tests and I dont have to make assumptions about the password which is manually set when the development database is created  (see run.sh script)

**Comment by afausti** [comment]
@jhoblitt I am glad that you were able to test the API even with little instructions, thank! 

I think @jmatt can help us in reviewing the django part.

File: squash/dashboard/serializers.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        model = Job
+        fields = ('name', 'build', 'runtime',
+                  'url', 'status', 'measurements', 'links',)
+
+    """ Override the create method to create nested objects
+    from request data
+    """
+
+    def create(self, data):
+        measurements = data.pop('measurements')
```
Comment (jmatt) [inline squash/dashboard/serializers.py]: Measurements at this point are not correct.

```python
logging.debug(self.context['request']._request.body)
```
logging.debug(self.context['request']._request.body)

It looks like this:

```
2016-04-25 16:58:11,917 root-DEBUG [/Users/jmatt/dev/lsst/sqre/DM-5745/qa-dashboard/squash/dashboard/serializers.py 64] b'measurements=value&measurements=metric&url=https%3A%2F%2Fci.lsst.codes%2Fjob%2Fci_cfht%2F1%2F&status=0&name=ci_cfht&build=1&runtime=2016-04-24T19%3A26%3A12.561564Z'
```

Which does not include the actual value and metric value. I think you need to pass it as json. Which can be done like this (for testing purposes):

```python
resp = requests.post(api['job'], json=job, auth=(TEST_USER, TEST_PASSWD))
```

Comment (jhoblitt) [inline squash/dashboard/serializers.py]: perhaps there is a version difference at play here?  It seems to be working for me:

```sh
$  curl -i --user sqre-admin:foo -Happlication/json" -H "Content-Type: application/json" -X GET http://localhost:8000/api/job/
HTTP/1.0 200 OK
Date: Mon, 25 Apr 2016 19:32:01 GMT
Server: WSGIServer/0.1 Python/2.7.5
Vary: Accept
X-Frame-Options: SAMEORIGIN
Content-Type: application/json
Allow: GET, POST, HEAD, OPTIONS

{"count":2,"next":null,"previous":null,"results":[{"name":"ci_cfht","build":"1","runtime":"2016-04-25T19:29:36.901849Z","url":"https://ci.lsst.codes/job/ci_cfht/1/","status":0,"measurements":[],"links":{"self":"http://localhost:8000/api/job/1/"}},{"name":"ci_cfht","build":"1","runtime":"2016-04-25T19:29:45.870848Z","url":"https://ci.lsst.codes/job/ci_cfht/1/","status":0,"measurements":[],"links":{"self":"http://localhost:8000/api/job/2/"}}]} 
$  curl -i --user sqre-admin:foo -H "Accept: application/json" -H "Content-Type: application/json" -X GET http://localhost:8000/api/metric/
HTTP/1.0 200 OK
Date: Mon, 25 Apr 2016 19:32:18 GMT
Server: WSGIServer/0.1 Python/2.7.5
Vary: Accept
X-Frame-Options: SAMEORIGIN
Content-Type: application/json
Allow: GET, POST, HEAD, OPTIONS

{"count":1,"next":null,"previous":null,"results":[{"metric":"test1","description":"Test metric insertion","units":"test","condition":"<","minimum":8.0,"design":5.0,"stretch":3.0,"user":5.0,"links":{"self":"http://localhost:8000/api/metric/test1/"}}]}
```

Comment (jmatt) [inline squash/dashboard/serializers.py]: Yeah sorry about that. It was out of context. This was specifically to address @afausti problem in [this JIRA issue comment](https://jira.lsstcorp.org/browse/DM-5745?focusedCommentId=47227&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-47227).

**Pull request merged**